README.md
# connection limit middlware for koa

[![Build Status](https://travis-ci.org/vicanso/koa-connection-limit.svg?style=flat-square)](https://travis-ci.org/vicanso/koa-connection-limit)
[![Coverage Status](https://img.shields.io/coveralls/vicanso/koa-connection-limit/master.svg?style=flat)](https://coveralls.io/r/vicanso/koa-connection-limit?branch=master)
[![npm](http://img.shields.io/npm/v/koa-connection-limit.svg?style=flat-square)](https://www.npmjs.org/package/koa-connection-limit)
[![Github Releases](https://img.shields.io/npm/dm/koa-connection-limit.svg?style=flat-square)](https://github.com/vicanso/koa-connection-limit)

## Installation

```bash
$ npm install koa-connection-limit
```

## API

```js
const koa = require('koa');
const koaConnectionLimit = require('koa-connection-limit');
const app = koa();

app.use(koaConnectionLimit({
  mid: 5,
  high: 10,
  throwError: false,
  pass: (ctx) => {
    return false;
  }
}, function (status) {
  // status: low, mid, high
  console.info(status);
}));
```
### Options

- `mid` mid connection limit count
- `high` high connection limit count
- `throwError` when `true` or `undefined`, the connection count reach high limit count, it will throw error
- `pass` if the function return true, the request will be ingore of limit
- `err` the custom define error, optional

### onChange

when status change, the function will be triggered


## Example

```js
'use strict';
const Koa = require('koa');
const app = new Koa();
const koaConnectionLimit = require('koa-connection-limit');

// logger

app.use((ctx, next) => {
  const start = new Date;
  return next().then(() => {
    const ms = new Date - start;
    console.log(`${ctx.method} ${ctx.url} - ${ms}ms`);
  });
});


app.use(koaConnectionLimit({
  high: 2,
  mid: 1,
  throwError: false,
  pass: (ctx) => {
    return ctx.url === '/no-limit';
  },
}, function changeStatus(status) {
  console.info(status);
}));


app.use((ctx, next) => {
  const delay = new Promise(function(resolve, reject) {
    setTimeout(resolve, 3000);
  });
  return delay.then(next);
});

// response

app.use(ctx => {
  ctx.body = 'Hello World';
});

app.listen(3000);
```


## License

MIT


History.md
# 2.1.0 (2017-03-29)
 * Remove the `util` node module
 * Support `pass` function to pass the limit
 * Support customize error


test/limit.js
'use strict';
const assert = require('assert');
const Koa = require('koa');
const koaConnectionLimit = require('../lib/limit');
const request = require('supertest');

describe('koa-connection-limit', function() {
  it('should throw error when mid or high is invalid', function(done) {
    // mid is not a number
    try {
      koaConnectionLimit({
        mid: 'a',
      });
    } catch (err) {
      assert.equal('mid and high must be a number', err.message);
    }
    // high is not a number
    try {
      koaConnectionLimit({
        high: 'a',
      });
    } catch (err) {
      assert.equal('mid and high must be a number', err.message);
    }
    // mid is lt 1
    try {
      koaConnectionLimit({
        mid: -1,
        high: 10,
      });
    } catch (err) {
      assert.equal('mid and high must gt 0', err.message);
    }
    // high is lt 1
    try {
      koaConnectionLimit({
        mid: 10,
        high: -1,
      });
    } catch (err) {
      assert.equal('mid and high must gt 0', err.message);
    }
    // mid gt high
    try {
      koaConnectionLimit({
        mid: 10,
        high: 2,
      });
    } catch (err) {
      assert.equal('high must be gt mid', err.message);
    }

    done();
  });

  it('should set connection limit successful', function(done) {
    const app = new Koa();
    app.use(koaConnectionLimit({
      mid: 1,
      high: 2,
    }));

    app.use((ctx) => {
      return new Promise(function(resolve) {
        setTimeout(() => {
          ctx.body = 'OK';
          resolve();
        }, 500);
      });
    });
    const total = 3;
    const server = app.listen();
    const statusList = [429, 200, 200];
    let finishCount = 0;
    const finish = (err, res) => {
      assert.equal(res.status, statusList[finishCount]);
      finishCount++;
      if (finishCount === total) {
        done();
      }
    };
    for (let i = 0; i < total; i++) {
      setTimeout(() => {
        request(server)
          .get('/')
          .end(finish);
      }, i * 10);
    }
  });

	it('should pass connection limit successful', function(done) {
    const app = new Koa();
    app.use(koaConnectionLimit({
      mid: 1,
      high: 2,
			pass: () => true,
    }));

    app.use((ctx) => {
      return new Promise(function(resolve) {
        setTimeout(() => {
          ctx.body = 'OK';
          resolve();
        }, 500);
      });
    });
    const total = 3;
    const server = app.listen();
    const statusList = [200, 200, 200];
    let finishCount = 0;
    const finish = (err, res) => {
      assert.equal(res.status, statusList[finishCount]);
      finishCount++;
      if (finishCount === total) {
        done();
      }
    };
    for (let i = 0; i < total; i++) {
      setTimeout(() => {
        request(server)
          .get('/')
          .end(finish);
      }, i * 10);
    }
  });

  it('should onChange event successful', function(done) {
    const app = new Koa();
    app.use(koaConnectionLimit((status) => {
      assert.equal(status, 'low');
    }));

    request(app.listen())
      .get('/')
      .expect(404, done);
  });


  it('should onChange event successful when throw an error', function(done) {
    const app = new Koa();
    app.use(koaConnectionLimit((status) => {
      assert.equal(status, 'low');
    }));
    app.use(ctx => i.j = 0);
    request(app.listen())
      .get('/')
      .expect(500, done);
  });

});


gulpfile.js
var gulp = require('gulp');
var jshint = require('gulp-jshint');
var mocha = require('gulp-mocha');
var istanbul = require('gulp-istanbul');


gulp.task('jshint', function () {
  return gulp.src('./lib/*.js')
    .pipe(jshint({
      predef: ['require', 'module'],
      node: true,
      esnext: true
    }))
    .pipe(jshint.reporter('default'));
});

gulp.task('test', function (cb) {
  gulp.src(['./lib/*.js'])
    .pipe(istanbul())
    .pipe(istanbul.hookRequire())
    .on('finish', function () {
      gulp.src(['./test/*.js'])
        .pipe(mocha({
          reporter: 'spec'
        }))
        .pipe(istanbul.writeReports({
          reporters: ['text', 'text-summary', 'html']
        }))
        .pipe(istanbul.enforceThresholds({
          thresholds: {
            global: 90
          }
        })).on('end', cb);
    });
});

gulp.task('default', ['jshint', 'test']);


lib/limit.js
'use strict';

const debug = require('debug')('koa-connection-limit');
/* istanbul ignore next */
function noop() {}


function isFunction(value) {
  return typeof value === 'function';
}

function isObject(value) {
  return typeof value === 'object';
}

function isUndefined(value) {
  return value === undefined;
}

function get(arr, filter, defaultValue) {
  let result;
  arr.forEach((tmp) => {
    if (tmp && filter(tmp)) {
      result = tmp;
    }
  });
  return result || defaultValue;
}

/**
 * [limit description]
 * @param  {[type]} options  [description]
 * @param  {[type]} onChange [description]
 * @return {[type]}          [description]
 */
function limit() {
  const args = Array.from(arguments);
  const onChange = get(args, isFunction, noop);
  const error = new Error('Too Many Requests');
  error.status = 429;
  const options = Object.assign({
    err: error,
    mid: 100,
    high: 500,
  }, get(args, isObject));
  debug('limit options:', options);
  const mid = parseInt(options.mid, 10);
  const high = parseInt(options.high, 10);
  if (isNaN(mid) || isNaN((high))) {
    throw new Error('mid and high must be a number');
  }
  if (mid <= 0 || high <= 0) {
    throw new Error('mid and high must gt 0');
  }
  if (mid >= high) {
    throw new Error('high must be gt mid');
  }
  let handlingReqTotal = 0;
  let status = '';
  let throwError = options.throwError;
  /* istanbul ignore else */
  if (isUndefined(throwError)) {
    throwError = true;
  }
  const pass = options.pass || noop;
  return (ctx, next) => {
    function checkStatus() {
      if (handlingReqTotal > high) {
        /* istanbul ignore else */
        if (status !== 'high') {
          status = 'high';
          onChange(status);
        }
        /* istanbul ignore else */
        if (throwError) {
          throw options.err;
        }
      } else if (handlingReqTotal > mid) {
        /* istanbul ignore else */
        if (status !== 'mid') {
          status = 'mid';
          onChange(status);
        }
      } /* istanbul ignore else */ else if (status !== 'low') {
        status = 'low';
        onChange(status);
      }
    }

    function done(err) {
      handlingReqTotal -= 1;
      checkStatus();
      if (err) {
        throw err;
      }
    }
    if (pass(ctx)) {
      return next();
    }

    handlingReqTotal += 1;
    checkStatus();
    return next().then(done, done);
  };
}

module.exports = limit;
